src/Microsoft.DotNet.Cli.Utils/CommandResolver.cs

```diff
@@ -0,0 +1,191 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using NuGet.Frameworks;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.DotNet.ProjectModel.Graph;
+using NuGet.Packaging;
+
+namespace Microsoft.DotNet.Cli.Utils
+{
+    internal static class CommandResolver
+    {
+        public static CommandSpec TryResolveCommandSpec(string commandName, string args, NuGetFramework framework = null)
+        {
+            return ResolveFromRootedCommand(commandName, args) ??
+                   ResolveFromProjectDependencies(commandName, args, framework) ??
+                   ResolveFromProjectTools(commandName, args) ??
+                   ResolveFromPath(commandName, args);
+        }
+
+        private static CommandSpec ResolveFromPath(string commandName, string args)
+        {
+            var commandPath = Env.GetCommandPath(commandName);
+
+            if (commandPath == null) return null;
+            
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) &&
+                Path.GetExtension(commandPath).Equals(".cmd", StringComparison.OrdinalIgnoreCase))
+            {
+                var preferredCommandPath = Env.GetCommandPath(commandName, ".exe");
+
+                if (preferredCommandPath != null)
+                {
+                    commandPath = Environment.GetEnvironmentVariable("ComSpec"); 
+
+                    args = $"/S /C \"\"{preferredCommandPath}\" {args}\"";
+                }
+            }
+
+            return new CommandSpec(commandPath, args, CommandResolutionStrategy.Path);
+        }
+
+        private static CommandSpec ResolveFromRootedCommand(string commandName, string args)
+        {
+            if (Path.IsPathRooted(commandName))
+            {
+                return new CommandSpec(commandName, args, CommandResolutionStrategy.Path);
+            }
+
+            return null;
+        }
+
+        public static CommandSpec ResolveFromProjectDependencies(string commandName, string args,
+            NuGetFramework framework)
+        {
+            if (framework == null) return null;
+
+            var projectContext = GetProjectContext(framework);
+
+            if (projectContext == null) return null;
+
+            var commandPackage = GetCommandPackage(projectContext, commandName);
+
+            if (commandPackage == null) return null;
+
+            var depsPath = projectContext.GetDepsPath(Constants.DefaultConfiguration);
+
+            return ConfigureCommandFromPackage(commandName, args, commandPackage, projectContext, depsPath);
+        }
+
+        private static ProjectContext GetProjectContext(NuGetFramework framework)
+        {
+            var projectRootPath = Directory.GetCurrentDirectory();
```

Comment:
Isn't it risky to rely on the current directory? Can there be a situation where the current directory is different from where the desired project.json is?

TryResolveCommandSpec could receive more context on where the lookup should start. Or maybe even the ProjectContext in whose context this command is required.